SUIT draft v4 support

kaspar030 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR contains the combined commits for SUIT draft v4 support, against "master-plus-dependencies".

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

![suit-pr](https://user-images.githubusercontent.com/4679640/60731089-e7faa100-9f46-11e9-818f-d53f7e5f5aec.png)


closes #8, #19 and #20

basilfx committed:
gitignore: add clangd compile flags

bergzand committed:
gnrc_rpl: remove DEBUG from header files

This conflicts with the DEBUG setting in C files that include these
headers making it impossible to enable the DEBUG setting in these C
files.

mtausig committed:
crypto: aes_init(): Fail correctly when called with bad key length

A proper error code is returned if a key with unsupported (either by the implementation or the AES algorithm) length is passed to aes_init.
This fixes Issue #10175

mtausig committed:
crypto: Call AES tests with real key length

mtausig committed:
crypto: Fix code style

mtausig committed:
crypto: Add tests for aes_init return value on bad key length

jcarrano committed:
tests/shell: use tuple instead of dictionary in test scripts.

Python dictionaries are not guaranteed to be ordered until version
3.7. In 3.6 they are ordered too, but that is an implementation
detail. riotdocker seems to be using 3.5.

As it stands now, it would not be a problem if the test commands
are run in a random order, except that:

- It would result in non-reproduceable tests.
- It hinders testing other functionality, such as exiting the shell.

jcarrano committed:
tests/shell: Test cancelling a line with ctrl-C.

Send garbage, cancel it and issue a valid command. No errors should
ocurr.

gschorcht committed:
cpu/esp32: esp_wifi compile configuration tuned

Downsized numbers of dynamic send and receive buffers.

gschorcht committed:
cpu/esp*: shorter netdev task names

gschorcht committed:
cpu/esp_common: add/fix comments and messages

gschorcht committed:
cpu/esp_common: set WiFi config storage type

gschorcht committed:
cpu/esp32: cleanup comments/messages in esp_wifi

gschorcht committed:
cpu/esp32: set WiFi config storage type

gschorcht committed:
cpu/esp_common: cleanup of compile time config

gschorcht committed:
cpu/esp_common: doc fixes

gschorcht committed:
cpu/esp*: changes to use esp_wifi and esp_now

miri64 committed:
gnrc_sixlowpan_frag: initial import of the VRB

VRB = virtual reassembly buffer

jcarrano committed:
sys/shell: cancel current line on CTRL-C.

CTRL-C cancels the current line, similar to how getty works.

This is useful if one is using a dumb terminal to communicate with
a node, as it saves having to repeatedly type backspace to discard the
current line. It also helps when connecting to an already running node,
as one does not know what is on the line buffer, the safest thing to do
is to begin by sending a ctrl-C.

This is a suggestion of @benemorius.

miri64 committed:
tests/gnrc_ndp: enhance coverage

miri64 committed:
tests: add unittests for gnrc_sixlowpan_frag_vrb

miri64 committed:
gnrc_sixlowpan_frag: include VRB in normal GC

cladmi committed:
test_utils_interactive_sync: add a helper for synchronizing tests

Add an implementation that waits for 's' to print 'START' and return.
If 'r' is given is prints 'READY' to allow querying for state.

The help and answered string have to be different to not match the other.
Using puts/getchar was smaller than using `stdio_read/stdio_write` on the
example I tested with `esp32`.

cladmi committed:
testrunner/utils: add helper for test_utils_interactive_sync

Add helper to do the synchronisation.

aabadie committed:
boards/stm32*: remove useless RTC_NUMOF defines

miri64 committed:
gnrc_netif: allow for wait of minimum time between sends

benpicco committed:
cpu/sam0_common: add hwrng driver

aabadie committed:
pkg/nimble: adapt to nrf51 family

cgundogan committed:
shell/ccn-lite: type fixes

aabadie committed:
stm32-common/spi: add customizable gpio modes for spi pins

fjmolinas committed:
stm32l1/vendor: add v2.3.0 vendor files

fjmolinas committed:
stm321l1/vendor: migrate to new v2.3.0 vendor files

- remove old header files
- fix new DMA header file macro definitions
- remove old cpu type groups (STM32L1XX_MD, STM32L1XX_MDP,
  STM32L1XX_HD, STM32L1XX_XL)

aabadie committed:
boards/stm32: provide support custom spi pins mode for lora boards

This minimize a lot power consumption when going to sleep after the spi has been used

aabadie committed:
drivers/spi: extend API with updatable gpio mode on pins

aabadie committed:
drivers/sx127x: let sx127x driver manage spi pins GPIO modes

miri64 committed:
posix_socket: initialize uninitialized fields

aabadie committed:
tests/log_printfnoformat: add test application & script

JulianHolzwarth committed:
xtimer/xtimer.c: ensure modifing mutex is done with interrupt disabled

if xtimer_set spins the callback is executed in the thread context.

comment to explain irq_disable
and when this line could be removed
(when xtimer stops executing the callback funtion from thread context)

aabadie committed:
native: don't use NATIVEINCLUDES in board and board/drivers

haukepetersen committed:
boards/common/nrf51: add alternative xtimer mapping

JulianHolzwarth committed:
xtimer/xtimer.c: _mutex_timeout improvements

Only yields and change threads status when thread was removed from mutex list.

fjmolinas committed:
boards/stm32f429i-disc1: add riotboot

fjmolinas committed:
stm32f4/Makefile.include: add riotboot requirements

- Stm32f4 use sectors instead of pages. They go from 16 KB to 128KB.
  The bootloader will use the first sector(16Kb). Slots must start
  at the begining of a sector to not overlap.
- Minimum required RIOBOOT_HDR_LEN or stm32f4 is 0x200
  to respect vector table alignment

fjmolinas committed:
boards/nucleo-f446re: add riotboot

benpicco committed:
cpu: saml21: don't silently ignore missing makefile

-include will throw no error if sam0_common/Makefile.features does not exist.
This may not have been intentional as none of the other sam0 implementations
do this.

Replace it with a normal include.

haukepetersen committed:
examples/emcute_mqttsn: allow override of EMCUTE_ID

haukepetersen committed:
tests: add Nordic SoftDevice test app

aabadie committed:
tools/flake8: exclude suit_manifest_encoder_04 script

This script is copied as is from an external repository:
https://github.com/ARMmbed/suit-manifest-generator/blob/ex-gen-moran-suit-manifest-04/example-generator-moran-suit-manifest-04/make-examples.py

JulianHolzwarth committed:
xtimer/xtimer.c: sched_switch instead of yield

because of pr #11759: not all boards check for is_in_irq when thread_yield_higher

kaspar030 committed:
suit: add SUIT draft v4 manifest tools

This commit adds support tools used by the SUIT firmware upgrade module.

Co-authored-by: Alexandre Abadie <alexandre.abadie@inria.fr>
Co-authored-by: Koen Zandberg <koen@bergzand.net>
Co-authored-by: Francisco Molina <femolina@uc.cl>

kaspar030 committed:
sys/suit: add SUIT draft v4 firmware upgrade module

This commit adds a sys module implementing SUIT draft v4 compatible
firmware updates.

Co-authored-by: Alexandre Abadie <alexandre.abadie@inria.fr>
Co-authored-by: Koen Zandberg <koen@bergzand.net>
Co-authored-by: Francisco Molina <femolina@uc.cl>

kaspar030 committed:
examples/suit_update: add SUIT draft v4 example & test

This commit adds an example application showcasing SUIT draft v4
firmware updates.

It includes a test script suitable for local or CI testing.

Co-authored-by: Alexandre Abadie <alexandre.abadie@inria.fr>
Co-authored-by: Koen Zandberg <koen@bergzand.net>
Co-authored-by: Francisco Molina <femolina@uc.cl>

fjmolinas commented on examples/suit_update/Makefile:
```diff
@@ -0,0 +1,117 @@
+  #
+  #... in another shell and keep it running.
+  export TAP ?= riot0
+  TERMPROG = $(RIOTTOOLS)/ethos/ethos
```
There is an issue where if I kill the setup script before the terminal the tap interface doesn't get cleaned.

```
sudo dist/tools/ethos/setup_network.sh riot0 2001:db8::/64
net.ipv6.conf.riot0.forwarding = 1
net.ipv6.conf.riot0.accept_ra = 0
Joining IPv6 multicast group...
entering loop...
got packet from fe80::2 port 12345
got packet from fe80::2 port 12345
got packet from fe80::2 port 12345
got packet from fe80::2 port 12345
got packet from fe80::2 port 12345
^CCleaning up...
ioctl(TUNSETIFF): Device or resource busy

```

kaspar030 replied:
uff, there's not much to do about this. the tap interface is still open.
We *could* hack something with lsof, but that would be ugly and non-portable...
Alternatively we could try removing the tap in a loop, like

```
while ip tuntab mode tap del $tap ; do
  echo "tap busy, ethos still running?"
  sleep 1
done
```

Do you think we need to solve this before PRing?

fjmolinas replied:
No, lets ~not~solve it afterwards

fjmolinas commented on examples/suit_update/README.md:
```diff
@@ -0,0 +1,548 @@
+If the workflow for updating using ethos is successful, you can try doing the
+same over "real" network interfaces, by updating a node that is connected
+wirelessly with a border router in between.
+
```
I had to add this to make it work without automatic script:

```
+From another terminal on the host, add a routable address on the host `riot0`
+interface:
+
+    $ sudo ip address add 2001:db8::1/128 dev riot0
+
+In another terminal, run:
+
+    $ BOARD=samr21-xpro make -C examples/suit_update/ term
```

fjmolinas commented on makefiles/suit.v4.inc.mk:
```diff
@@ -0,0 +1,106 @@
+suit/delkeys:
+	rm -f $(SUIT_SEC) $(SUIT_PUB)
+
+#
```
I think this can be removed

kaspar030 replied:
removed

fjmolinas commented on sys/include/suit/coap.h:
```diff
@@ -0,0 +1,96 @@
+    }
+
+
+/* this is internal code that will go soon, thus skip Doxygen checks */
```
I think this needs a better justification.

kaspar030 replied:
added some prose, please check

fjmolinas replied:
Perfect

fjmolinas commented on sys/include/suit/v4/handlers.h:
```diff
@@ -0,0 +1,65 @@
+ * @return      ptr to handler function
+ * @return      NULL (if handler unavailable or key out of range)
+ */
+suit_manifest_handler_t suit_manifest_get_handler(int key);
```
This function is never used, the doxygen documentation can be used for the upper function.

kaspar030 replied:
fixed

fjmolinas commented on sys/include/suit/v4/handlers.h:
```diff
@@ -0,0 +1,65 @@
+ */
+typedef int (*suit_manifest_handler_t)(suit_v4_manifest_t *manifest, int key, CborValue *it);
+
+suit_manifest_handler_t suit_manifest_get_manifest_handler(int key);
```
missing doxygen

fjmolinas commented on sys/suit/coap.c:
```diff
@@ -0,0 +1,496 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
```
Missing doxygen headers and authorship.

fjmolinas reviewed (changes requested):
I performed the following tests:

- automatic tests
- wireless updates
- wired updates

I only have two functionality comments, one regarding script cleanup (see below).

The second is documentation changes, I made a [gist](https://gist.github.com/fjmolinas/6858e2452f3bd8be4893e1178231f5fd) to make applying it easier, its mainly except where I added a comment in the code. I also changed all `tap0` to `riot0` for consistency.

The rest of the changes regard documentation.

fjmolinas commented on examples/suit_update/tests/01-run.py:
```diff
@@ -0,0 +1,130 @@
+MANIFEST_TIMEOUT = 15
+
+USE_ETHOS = int(os.getenv("USE_ETHOS", "1"))
+TAP = os.getenv("TAP", "tap0")
```
Also need to change to riot0 if my documentation changes are applied.

kaspar030 replied:
done

fjmolinas commented:
@kaspar030  I don't get what happened but the commit history is ruined now many unrelated changes and commits.

kaspar030 commented:
> @kaspar030 I don't get what happened but the commit history is ruined now many unrelated changes and commits.

Uh, I updated the suit-pr-base branch incorporating the new dependency structure, and pushed it. But I didn't push it to future-proof-iot but to my github repo. This PR was thus diffed against the outdated suit-pr-base from before... Should be fixed.

kaspar030 commented:
>  I made a [gist](https://gist.github.com/fjmolinas/6858e2452f3bd8be4893e1178231f5fd) to make applying it easier

applied. you can also push fixups directly to the branch I think, just let me know so I fetch them before rebaseing. :)

kaspar030 commented:
I have an issue compiling the updated libcose :(

aabadie commented on dist/tools/suit_v4/gen_key.py:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python3
+
```
Maybe add a copyright ? Same comment for gen_manifest.py

aabadie commented on examples/suit_update/Makefile:
```diff
@@ -0,0 +1,117 @@
+# Comment this out to enable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+#DEVELHELP = 1
```
Should not be commented out according to the comment above.

kaspar030 commented:
guys, please just push fixups!

aabadie commented on examples/suit_update/Makefile:
```diff
@@ -0,0 +1,117 @@
+  #     $ cd dist/tools/ethos; sudo ./setup_network.sh riot0 2001:db8::0/64
+  #
+  #... in another shell and keep it running.
+  export TAP ?= riot0
```
is the export really needed ?

kaspar030 replied:
yes, for the test script

fjmolinas commented:
> applied. you can also push fixups directly to the branch I think, just let me know so I fetch them before rebasing. :)

Will do next time. I don't think you've pushed though.

> I have an issue compiling the updated libcose :(

what commi?

aabadie commented:
> guys, please just push fixups!

Ok, I'll add the missing copyrights this way :)

aabadie commented on examples/suit_update/tests/01-run.py:
```diff
@@ -0,0 +1,130 @@
+
+
+def start_aiocoap_fileserver():
+    tmpdirname = tempfile.TemporaryDirectory()
```
I would refactor the tmpdir a bit: add it as global variable, just under `TAP`, call it `_TMP_DIR` and initialize it directly with `tempfile.TemporaryDirectory()`.
Then it can be reused from within any function. This function would just return the aiocoap_process variable. `cleanup` just takes an argument and no need to declare tmpdir `global` from several places.

What do you think ?

kaspar030 replied:
> What do you think ?

done!

fjmolinas commented:
@kaspar030 I pushed the gist.

fjmolinas commented:
> I have an issue compiling the updated libcose :(

I now have the same issue...

I pushed the missing changes I requested.

aabadie committed:
boards/stm32: introduce common timer config with TIM2

aabadie committed:
boards/nucleo-f0/3/7: use common timer configuration

aabadie committed:
boards/stm32l0: use common timer configuration

aabadie committed:
boards/nucleo-l4*: use common timer configuration

aabadie committed:
boards/stm32: introduce common timer config with TIM5

aabadie committed:
boards/stm32f4: use common timer configuration

aabadie committed:
boards/ublox-c030-u201: use common timer configuration

This also requires to include the STM32 shared include directory

aabadie committed:
boards/lobaro-lorabox: use common timer configuration

aabadie committed:
boards/nucleo-l152re: use TIM5 common timer configuration

aabadie committed:
boards/nz32-sc151: use TIM5 common timer configuration

This requires including STM32 common include directoy as well

aabadie committed:
boards/stm32f769i-disco: deduplicate doxygen group definition

The group is already defined in the doc.txt file so no need to define it again in board.h

aabadie committed:
boards/esp8266: deduplicate board doxygen group definition

bergzand commented:
> I have an issue compiling the updated libcose :(

> I now have the same issue...

I was able to reproduce the issue. No idea what was the exact cause, but I suspect a compiler bug in combination with the multiple `case` statements [here](https://github.com/bergzand/libcose/blob/5442e1ec94b203414affe7d7f8f1238e10983c3d/src/cose_hdr.c#L57-L59). I reworked the function a bit and the latest libcose master should work again.

aabadie committed:
tools/doccheck: add check for multiple group definition

aabadie committed:
cpu/efm32: deduplicate cpu_efm32 group definition

aabadie committed:
sys/sha1: deduplicate sys_hashes_sha1 group definition

aabadie committed:
sys/pthread: deduplicate pthread group definition

aabadie committed:
pkg/lwip: cleanup doxygen documentation

- Deduplicate lwip group definition
- Fix reference to sock instead of conn

aabadie committed:
boards/common/esp: deduplicate common esp boards group definitions

aabadie committed:
boards/arduino-leonardo: remove duplicate doc group definition

aabadie committed:
core/sched_native: fix duplicate core_sched group definition

Introduce core_sched_native group and put this in group in parent
core_sched group

aabadie committed:
net/gnrc/sock: remove duplicate net_gnrc_sock group definition

aabadie committed:
net/gnrc/tcp: remove duplicate net_gnrc_tcp group definitions

aabadie committed:
boards/nucleo-*: deduplicate doxygen board group definitions

fjmolinas commented:
I'm still having issues with libcose, even with the fix. I had to add `CFLAGS += -Wno-unused-parameter` because of unused parameters in `cose_crypto_sign()` and then I get:

```
"make" -C /home/francisco/workspace/RIOT/examples/suit_update/bin/pkg/samr21-xpro/libcose/src/crypt -f /home/francisco/workspace/RIOT/pkg/libcose/Makefile.libcose_crypt
make[2]: *** No rule to make target 'tweetnacl.c', needed by '/home/francisco/workspace/RIOT/examples/suit_update/bin/samr21-xpro/libcose_crypt/tweetnacl.o'.  Stop.
Makefile:9: recipe for target 'all' failed
make[1]: *** [all] Error 2
/home/francisco/workspace/RIOT/examples/suit_update/../../Makefile.include:524: recipe for target 'pkg-build-libcose' failed
make: *** [pkg-build-libcose] Error 2
make: Leaving directory '/home/francisco/workspace/RIOT/examples/suit_update'

```

Is it working for you @kaspar030 ?

bergzand commented:
@fjmolinas Is there a `libcose_crypto_$cryptopkg` enabled? (there should be) I think `libcose_crypt_hacl` is the only one working at the moment as the monocypher and c25519 are missing a tiny bit of RIOT glue. I've phased out support for tweetnacl with libcose as it is too much trouble to maintain the glue.

kaspar030 commented:
I'm trying right now. Does the error go away if you "make clean" first? I had some issues with bad bin/ state before.

fjmolinas commented:
@bergzand There is:

`USEMODULE += libcose_crypt_tweetnacl` should it be `USEMODULE += libcose_crypt_hacl`?

bergzand commented:
> `USEMODULE += libcose_crypt_tweetnacl` should it be `USEMODULE += libcose_crypt_hacl`?

Yes. Current libcose master doesn't support tweetnacl anymore :/

fjmolinas commented:
> Yes. Current libcose master doesn't support tweetnacl anymore :/


Ok, that works, there are some missing api migrations on the suit part. Working on it!

kaspar030 commented:
> there are some missing api migrations on the suit part. Working on it!

Ok, cool!

kaspar030 commented:
I've force-pushed an update to suit-pr-base including a libcose bump with the latest fix, and a rebase of this PR. It includes a change to make libcose use hacl.

fjmolinas commented:
I've pushed to fixup commits:

- make the api changes according to latest libcose structure
-  re-add `riotboot_flashwrite_verify_sha256` from the headers it had been removed

Everything is working for me now.

kaspar030 commented:
> Everything is working for me now.

Confirmed here. I'll try on CI.

fjmolinas commented on examples/suit_update/tests/01-run.py:
```diff
@@ -0,0 +1,128 @@
+        )
+        target_slot = int(child.match.group(1))
+        # Wait for update to complete
+        child.expect_exact(
```
I was thinking of changing the way it waits for the update to complete to this:

```
def wait_for_update(child):
    try:
        while True:
            child.expect(r"riotboot_flashwrite: processing bytes (\d+)-(\d+)",
                         timeout=TIMEOUT)
    except pexpect.TIMEOUT:
        child.expect_exact(
            "riotboot_flashwrite: riotboot flashing completed successfully",
            timeout=TIMEOUT)
        print(child.after)
```

That way it doesnt depend on a very long timeout that can vary according to the board/cpu.

fjmolinas replied:
It works for me, any reason why it wouldn't in the CI or why this is a bad idea? @kaspar030

kaspar030 replied:
This way it'll always timeout at least once (after the update is complete). Isn't it possible to wait for two strings with pexpect?

kaspar030 replied:
See here: https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.expect

You can just pass a list.

How about:

```
while True:
            child.expect([r"riotboot_flashwrite: processing bytes (\d+)-(\d+)", "riotboot_flashwrite: riotboot flashing completed successfully"])
                         timeout=TIMEOUT)
```

kaspar030 replied:
well, need to check the result:

```
while True:
            index = child.expect([r"riotboot_flashwrite: processing bytes (\d+)-(\d+)",
                                  "riotboot_flashwrite: riotboot flashing completed successfully"])
                         timeout=TIMEOUT)
             if index == 1:
                  break
```

fjmolinas replied:
I did something like that:

```
def wait_for_update(child):
        return child.expect([r"riotboot_flashwrite: processing bytes (\d+)-(\d+)",
                             "riotboot_flashwrite: riotboot flashing "
                             "completed successfully"],
                             timeout=UPDATING_TIMEOUT)






# Wait for update to complete
while wait_for_update(child) == 0:
     pass
```

fjmolinas replied:
You ok with the change? Tested and ready to push!

kaspar030 replied:
yes, nice!

fjmolinas reviewed (approved):
Tests are passing. I say squash, run tests one more time in CI and open PR!

kaspar030 commented:
- squashed

fjmolinas commented:
@kaspar030 did it pass on nrf52?

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

fjmolinas committed:
cpu/cortexm_common: replace irq_restore by __set_PRIMASK for stm32l152re

- The __NOP() that was added in #8518 is now remooved.
- When DBG_STANDBY, DBG_STOP or DBG_SLEEP are set in DBG_CR a hardfault
  occurs on wakeup from sleep. This was first diagnosed in #8518. When
  enabled, a hardfault occured when returning from a branch to irq_restore()
  we avoid the call by inlining the function call. See #11830 for more
  details.

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

fjmolinas committed:
cpu/stm32_common: minimize consumption for STM32L1

- With this PR all GPIOs are set as AIN on start up.

Co-authored-by: Oleg Artamonov <oleg@unwds.com>

cladmi committed:
makefiles/buildtest: always execute 'buildtest' loop on host machine

This remove executing buildtest `for` loop in docker.

When building completely in docker, 'buildtest' would hide issues when
the host toolchain would be used when doing `make all` directly.
It has the consequence that it now starts a container for each
compilation which is slower.

The previous behavior can be reproduced by using

    BUILD_IN_DOCKER=1 make buildtest-indocker

A side effect is also that now `BUILDTEST_MAKE_REDIRECT` would work when
doing `buildtest` with docker.

fjmolinas committed:
stm32f7/Makefile.include: add riotboot requirements

- stm32f7 use sectors instead of pages, they go either from 16KB to
  128KB, or from 32KB to 25KB. Smaller sectors are at the begining of
  the flash. Slots must start at the begining of a sector to not overlap.
- Minimum required RIOBOOT_HDR_LEN or stm32f7 is 0x200
  to respect vector table alignment
- Add CPU_FLASH_BASE

fjmolinas committed:
boards: add riotboot for nucleo-f722ze/f746zg/f767zi

benpicco committed:
core: add ARRAY_SIZE macro

benpicco committed:
boards: make use of ARRAY_SIZE macro

benpicco committed:
tests: make use of ARRAY_SIZE macro

benpicco committed:
sys: make use of ARRAY_SIZE macro

benpicco committed:
drivers: make use of ARRAY_SIZE macro

benpicco committed:
examples: make use of ARRAY_SIZE macro

benpicco committed:
cpu: make use of ARRAY_SIZE macro

benpicco committed:
pkg: make use of ARRAY_SIZE macro

cladmi committed:
tests/cond_order: use test_utils_interactive_sync

Use test_utils_interactive_sync for synchronizing some case treat
the output before `reset` as the start of the test,
which fails for some boards/configurations.

benpicco committed:
examples: fix redundant assignment in TinyDTLS example

keestux committed:
core: make use of ARRAY_SIZE macro

Commit:
board/saml10-xpro: update implementation status in doc

Commit:
board/saml11-xpro: update implementation status in doc

Co-Authored-By: benpicco <benpicco@googlemail.com>

fjmolinas committed:
stm32f2/Makefile.include: add riotboot requirements

- stm32f2 uses sectors instead of pages, they go from 16KB to
  128KB. Smaller sectors are at the begining of the flash. Slots
  must start at the begining of a sector to not overlap.
- Minimum required RIOBOOT_HDR_LEN or stm32f2 is 0x200
  to respect vector table alignment

fjmolinas committed:
boards/nucleo-f207zg: add riotboot

miri64 committed:
github: add configuration for stale-bot

benpicco committed:
sys/shell: remove superfluous explicit cast

benpicco committed:
sys/shell: reduce scope of variable 'entry'

gschorcht committed:
sys/irq_handler: single interrupt handler thread

Single thread for handling interrupts that may trigger blocking  functions and therefore may only be called in thread context.

gschorcht committed:
tests/sys_irq_handler: test for sys/irq_handler

Test application for sys/irq_handler.

fjmolinas committed:
boards: use of default_i2c_cfg for nrf52 based boards

gschorcht committed:
Merge pull request #11239 from JulianHolzwarth/esp32/freertos/fix/semaphoreTake

cpu/esp32/freertos: fix semaphore take

gschorcht committed:
cpu/esp32: lwIP Ethernet address option length

The option value length of Ethernet addresses can be more than 6 byte in lwIP. Therefore, the max_len parameter is check to be greater than or equal to ETHERNET_ADDR_LEN.

gschorcht committed:
cpu/esp32: _esp_wifi_dev exposed for lwIP

To be able to access the single esp_wifi network device from lwIP adaptation layer, static keyword was removed from esp_wifi_dev variable.

kb2ma committed:
Merge pull request #11024 from kb2ma/coap/add_block_helpers

net/nanocoap: Buffer API Block helper functions

gschorcht committed:
pkg/lwip: add esp_wifi network device

The changes allow to use an esp_wifi network device of ESP32 with lwIP.

roberthartung committed:
cpu/atmega_common: implements pseudomodule-based pin change interrupts

roberthartung committed:
cpu/atmega1281: pin change interrupts

roberthartung committed:
cpu/atmega32u4: pin change interrupts

roberthartung committed:
cpu/atmega1284p: pin change interrupts

roberthartung committed:
cpu/atmega2560: pin change interrupts

roberthartung committed:
boards: adds atmega pin change interrupt to doc

roberthartung committed:
boards/arduino-mega2560: restrics usable pin change interrupts

roberthartung committed:
cpu/atmega256rfr2: pin change interrupts

MrKevinWeiss committed:
Merge pull request #10929 from gschorcht/esp_now_wifi_cleanup

cpu/esp*: esp_now and esp_wifi cleanup

haukepetersen committed:
pkg/nimble: bump version to 997dad8

haukepetersen committed:
tests/nimble_l2cap: adjust buffer sizes

miri64 committed:
Merge pull request #11949 from haukepetersen/opt_nimble_ver997dad8

pkg/nimble: bump version to 997dad8

haukepetersen committed:
Merge pull request #11576 from miri64/posix_socket/fix/fd-perms

posix_sockets: set read/write permissions for socket fd

haukepetersen committed:
examples/emcute: fix typo in README

PeterKietzmann committed:
pkg/ccn-lite: bump version

miri64 committed:
Merge pull request #11747 from haukepetersen/opt_emcute_nodeidoverride

examples/emcute_mqttsn: allow override of EMCUTE_ID

kb2ma committed:
Merge pull request #11784 from miri64/gcoap/fix/ack-variance-0

gcoap: don't call random_uint32_range() when COAP_ACK_VARIANCE=0

kb2ma committed:
Merge pull request #11787 from miri64/gcoap/enh/backoff-config

gcoap: allow for retransmission backoff to be turned off

cladmi committed:
compile_and_test_for_board: add --with-test-only

Add an option to only compile applications that have available test.

The target usage is when running hardware tests to save time by only
compiling what will be tested.

cladmi committed:
compile_and_test_for_board: add logger info for 'has_test'

Add a logger info printing if the application has a test.

maribu committed:
Merge pull request #11122 from ibr-cm/hartung/atmega-pin-change-interrupt-pseudomodules

cpu/atmega_common: pseudomodule-based pin change interrupt implementation

fjmolinas committed:
Merge pull request #11953 from cladmi/pr/tests/compile_and_test_for_board/only_compile_testable_applications

 compile_and_test_for_board: add --with-test-only

benpicco committed:
Merge pull request #11647 from benpicco/sam0-hwrng

cpu/sam0_common: add hwrng driver

gschorcht committed:
cpu/esp32: fix of the return code in esp_wifi_send

fjmolinas committed:
cpu/nrf5x_common: add flashpage_raw_support

fjmolinas committed:
cpu/efm32: add flashpage_raw

PeterKietzmann committed:
pkg/ccn-lite: suppress llvm error

miri64 committed:
Merge pull request #11952 from PeterKietzmann/pr_ccnl_update_and_bump

pkg/ccn-lite: bump version and adapt shell functions accordingly

aabadie committed:
Merge pull request #11919 from fjmolinas/pr_fix_11820_2

cpu/cortexm_common: replace irq_restore by __set_PRIMASK for stm32l152re

aabadie committed:
Merge pull request #11897 from fjmolinas/pr_stm32_common_define_alignment

cpu/stm32_common: uniformize define alignments

fjmolinas committed:
Merge pull request #11345 from aabadie/pr/boards/stm32_rtc_define_remove

boards/stm32*: remove useless RTC_NUMOF defines

haukepetersen committed:
Merge pull request #11797 from haukepetersen/add_test_softdevice

tests: add Nordic SoftDevice test app

maribu committed:
boards/{blue,black}pill: Fixed flashing

A new blue pill variant with only 32 KiB of flash (secretly coming with
64 KiB flash) instead of the use 64 KiB flash (secretly coming with 128 KiB)
is not compatible with `make flash`. This commit changes the OpenOCD config
so that both variants can be flashed.

haukepetersen committed:
tests/nimble_l2cap: adjust buffer sizes for nrf51

aabadie committed:
boards/nrf51: provide ble_nimble feature

fjmolinas committed:
tests/pkg_qdsa: increase timeout

- Some boards take longer than the default timeout to finish the test,
  e.g. nucleo-l152re

aabadie committed:
Merge pull request #11489 from fjmolinas/pr-update_stm32l1-cmsis

stm32l1/vendor: update vendor files to v2.3.0

fjmolinas committed:
Merge pull request #11812 from aabadie/pr/boards/nucleo_doc_cleanup

boards/nucleo*: move doxygen group definition to doc.txt

aabadie committed:
Merge pull request #11962 from fjmolinas/pr_nucleol152re_qdsa

tests/pkg_qdsa: increase timeout

aabadie committed:
Merge pull request #11939 from fjmolinas/pr_add_nrf52_i2c_cfg

boards/common/nrf52: add common nrf52 i2c config

Commit:
credman: add documentation for enum credman_type_t

fjmolinas committed:
Merge pull request #11542 from aabadie/stm32_spi_gpio_mode

stm32-common/spi: allow custom pin modes on spi to minimize power consumption

MrKevinWeiss committed:
Merge pull request #11959 from maribu/bluepill

boards/{blue,black}pill: Fixed flashing

MichelRottleuthner committed:
Merge pull request #11964 from gschorcht/cpu/esp32/esp_wifi/fix_send_return_value

cpu/esp32: fix of the return code in esp_wifi_send

benpicco committed:
sys/arduino: reduce scope of adc_line_state

This makes cppcheck happy.

aabadie committed:
drivers/periph_wdt: fix inline callback example

Function was renamed to wdt_setup_reboot_with_callback

aabadie committed:
drivers/periph_wdt: cleanup normal mode inline example

Remove a useless blankline

gschorcht committed:
cpu/esp32: fix unknown symbols for unused functions

Fixes the problem that the compilation of an applications can throw unknown symbol errors for functions that aren't use at all. Thus, it is possible to remove the warning for unknown symbols and the compilation can abort if there are real unknown symbols.

gschorcht committed:
cpu/esp32: fix pthread_setcancelstate symbol problem

When linking an application, symbol pthread_setcancelstate is not known in standard C libraries, even if the pthread module is linked. This is because the pthread module is grouped with all other modules, but not with the default C libraries when they are added to LINK_FLAGS. Therefore, standard C libraries have to be added also to BASELIBS to group them with all other modules.

gschorcht committed:
cpu/esp32: fix multiple definition of putchar

When standard C libraries are added to BASELIBS to group them together with all other modules, there are multiple definitions for the putchar function. The one that is defined writing to the UART as standard output and the one that is provided by the standard C libraries. To solve this symbol conflict, putchar and getchar functions that use the UART as standard output/input are renamed to __wrap_putchar and __wrap_getchar and the linker options are extended by -Wl,-wrap option.

gschorcht committed:
cpu/esp32: optimizing compilation of subdirs

A number of subdirectories in cpu/esp32/vendor/esp-idf have to be compiled ony, when according modules are required by the application.

aabadie committed:
Merge pull request #11966 from aabadie/pr/drivers/periph_wdt_doc_fix

drivers/periph_wdt: small doxygen fixes

aabadie committed:
Merge pull request #11832 from fjmolinas/pr_optimize_pm_stm32l1

cpu/stm32l1: optimize power consumption

miri64 committed:
Merge pull request #11814 from aabadie/pr/doc/remove_duplicate_group_define

doc: remove duplicate definitions of Doxygen groups

aabadie committed:
Merge pull request #11572 from aabadie/pr/boards/native_fix_log

boards/native: fix overriden INCLUDES leading to build failure

gschorcht committed:
cpu/esp32: fix multiple definitions with esp_idf_heap

If module esp_idf_heap is used, the memory management functions _malloc_r, _realloc_r, _calloc_r and _free_r have to be overridden by wrapper functions to use the heap_* functions of module _esp_idf_heap. However, this can lead to multiple symbol errors for these functions for some applications. To solve this symbol conflict, _malloc_r, _realloc_r, _calloc_r and _free_r functions are renamed to __wrap_* and the linker options are extended by -Wl,-wrap option when module esp_idf_heap is used.

miri64 committed:
gnrc_ipv6: check validity of preconfigured source on send

If an address was pre-configured by the upper layer its validity is
currently ignored. It is neither checked if the address is on the
interface at all nor is it checked if it is valid.

This change provides a fix for that by checking both facts.

MichelRottleuthner committed:
Merge pull request #11967 from gschorcht/cpu/esp32/make/fix-undefined-symbols

cpu/esp32: cleanup to fix undefined symbols

aabadie committed:
Merge pull request #11963 from fjmolinas/pr_efm32_flashpage_raw

cpu/efm32: add flashpage_raw

fjmolinas committed:
Merge pull request #11809 from aabadie/pr/boards/stm32_timer_conf_common

boards/stm32: introduce common timer configurations and use them where possible

aabadie committed:
tools/suit_v4: remove click dependency from gen_manifest.py

gschorcht committed:
cpu/esp32: fix conditional linker options for esp_idf_heap

Module esp_idf_heap is enabled in cpu/esp32/Makefile.dep depending on other modules. Since cpu/esp32/Makefile.dep is read after cpu/esp32/Makefile.include, the conditional  definition of the linker options for the wrapper functions had to be moved from cpu/esp32/Makefile.include to cpu/esp32/Makefile.dep.

dylad committed:
cpu/saml1x: fix adc resolution issue

MichelRottleuthner committed:
Merge pull request #11971 from gschorcht/cpu/esp32/make/fix-conditional-linker-options

cpu/esp32: fix of conditional linker options for esp_idf_heap

gschorcht committed:
pkg/lwip: override TPCIP_THREAD_PRIO for ESP32

 To avoid priority conflicts with the WiFi hardware driver thread which has priority of 1, the default thread priority of lwIP's TCP/IP thread is decreased to 2.

aabadie committed:
tools/doccheck: optimize display of undefined groups

The printed variable is not computed if no undefined group is found

keestux committed:
Merge pull request #11865 from benpicco/ARRAYSIZE

add ARRAY_SIZE macro and make use of it

keestux committed:
dist/tools: add Coccinelle check for ARRAYSIZE

This Coccinelle script will warn when new code has the potential to use
the macro ARRAYSIZE instead of something like this
     sizeof(some_array) / sizeof(some_array[0])

miri64 committed:
Merge pull request #11946 from gschorcht/cpu/esp32/lwip_netdev

cpu/esp32: lwIP netdev

fjmolinas committed:
boards: remove unused XTIMER_SHOOT_EARLY defines

aabadie committed:
Merge pull request #11956 from fjmolinas/pr_nrf5x_flashpage_raw

cpu/nrf5x_common: add flashpage_raw_support

aabadie committed:
Merge pull request #11974 from fjmolinas/pr_xtimer_shoot_early

boards: remove unused XTIMER_SHOOT_EARLY defines

jcarrano committed:
Merge pull request #10271 from mtausig/aes_rv

Fix return value of aes_init for keys of incorrect length

haukepetersen committed:
Merge pull request #11463 from aabadie/nimble_nrf51

pkg/nimble: adapt to nrf51 family

PeterKietzmann committed:
Merge pull request #11972 from dylad/pr/saml1x/fix_adc

cpu/saml1x: fix adc resolution issue

miri64 committed:
Merge pull request #11965 from pokgak/credman-add-credential-type-docs

credman: add documentation for enum credman_type_t

haukepetersen committed:
Merge pull request #11000 from miri64/gnrc_sixlowpan_frag/new/vrb

 gnrc_sixlowpan_frag: initial import of the VRB

fjmolinas committed:
Merge pull request #11687 from cladmi/pr/tests/pkg_u8g2/enable_in_ci

tests/pkg_u8g2: run the test in CI

MrKevinWeiss committed:
Merge pull request #11970 from miri64/gnrc_ipv6/fix/check-source-for-send

gnrc_ipv6: check validity of preconfigured source on send

benpicco committed:
Merge pull request #11973 from keestux/add-arraysize-cocci

dist/tools: add Coccinelle check for ARRAYSIZE

benpicco committed:
Merge pull request #11892 from Einhornhool/saml1x-doc-update

cpu/saml1x: update periph status implementation in board doc.txt

miri64 committed:
Merge pull request #11575 from miri64/posix_socket/fix/init-uninit-fields

posix_socket: initialize uninitialized fields

aabadie committed:
cpu/atmega_common/gpio: use ARRAY_SIZE macro

aabadie committed:
Merge pull request #11682 from fjmolinas/pr_stm32f4_riotboot

cpu/stm32f4: add riotboot requirements

miri64 committed:
Merge pull request #11981 from aabadie/pr/cpu/atmega_common_array_size

cpu/atmega_common/gpio: use ARRAY_SIZE macro

miri64 committed:
Merge pull request #11813 from aabadie/pr/tools/doccheck_detect_duplicate_defgroup

tools/doccheck: extend script to also check for Doxygen groups defined multiple times

fjmolinas committed:
cpu/kinetis: re-enable HWRNG for k64f

- Fix SCGx used for RNGA since RNGA is not mapped
  to SCG3 for K64F.

fjmolinas committed:
Merge pull request #11837 from miri64/gnrc_netif/enh/inter-tx-wait

gnrc_netif: allow for wait of minimum time between sends

aabadie committed:
Merge pull request #11900 from fjmolinas/pr_stm32f2_riotboot_requirements

cpu/stm32f2: add riotboot requirements

fjmolinas committed:
Merge pull request #10987 from miri64/tests/enh/gnrc_ndp-coverage

tests/gnrc_ndp: enhance coverage

miri64 committed:
Merge pull request #11938 from miri64/shell_commands/enh/now-id

shell_commands/ping6: mix-in current time into ID

kaspar030 committed:
Merge pull request #11921 from miri64/github/enh/stale-bot

github: add configuration for stale-bot

aabadie committed:
Merge pull request #11899 from fjmolinas/pr_stm32f7_riotboot_requirements

cpu/stm32f7: add riotboot requirements

fjmolinas committed:
Merge pull request #11933 from miri64/shell/fix/check-sender

shell_commands: check ICMPv6 echo reply sender

jcarrano committed:
Merge pull request #11985 from fjmolinas/pr_k64f_hwrng

cpu/kinetis: enable HWRNG for k64f

fabian18 committed:
drivers/include/periph/eeprom: Changed uint8_t* to void* in API

gschorcht committed:
sys/auto_init/can: cleanup for STM32 CAN controller

Initialization code for module `can_stm32` is obsolete since the STM32 CAN controller driver isn't a module anymore.

jcarrano committed:
Merge pull request #11807 from JulianHolzwarth/pr/xtimer_mutex_lock_timeout/first_fix_improvements

sys/xtimer/xtimer.c: _mutex_timeout() cleanup

benpicco committed:
Merge pull request #11990 from fabian18/drivers/include/periph/eeprom-api-change

drivers/include/periph/eeprom: Changed uint8_t* to void* in API

cladmi committed:
murdock: nordic_softdevice_ble: softdevice.hex is a test input

When running tests using 'nordic_softdevice_ble', the 'softdevice.hex'
file must also be taken into account for the test hashing and be
uploaded to the separated murdock testing boards.

When listed as dependency from `test-input-hash` the file must have a
target, which he has not. The file is implicitly created when compiling
`ELFFILE` so declare it as order only dependency.
In practice `BASELIBS` could be enough or even `pkg-build-softdevice`
but I do not want to be depend that much on internals there.

MrKevinWeiss committed:
Merge pull request #11995 from cladmi/pr/murdock/softdevice

murdock: nordic_softdevice_ble: softdevice.hex is a test input

cladmi committed:
tests/xtimer_usleep: use test_utils_interactive_sync

Use test_utils_interactive_sync for synchronizing test instead of the
custom 'getchar' handling.

cladmi committed:
tests/libfixmath: explicitly import ARRAY_SIZE definition

ARRAY_SIZE is implicitly imported by "xtimer.h" which will be removed of
this file.

cladmi committed:
tests/libfixmath: use test_utils_interactive_sync

Replace the 'xtimer_sleep' hack to prevent flooding at startup to use
test_utils_interactive_sync for synchronizing.

cladmi committed:
tests/posix_time: use test_utils_interactive_sync

Use test_utils_interactive_sync for synchronizing test instead of the
custom 'getchar' handling.

fjmolinas committed:
Merge pull request #11968 from aabadie/pr/tests/printf_format

tests/log_printfnoformat: add test application & script

gschorcht committed:
cpu/atmega_common: wrappers to avoid preemtion

Memory management function like `malloc`, `calloc`, `realloc` and `free` must not be preempted when they operate on allocator structures. To avoid such a preemption, wrappers around these functions are used which simply disable all interrupts for the time of their execution.

gschorcht committed:
cpu/atmega_common: activate wrapper functions when linking

nmeum committed:
gnrc_pktbuf: fix typo (guaranteeed → guaranteed)

nmeum committed:
gnrc_tcp: Fix memory leak

fjmolinas committed:
Merge pull request #12000 from nmeum/pr/gnrc_pktbuf_mark_typo

gnrc_pktbuf: fix typo (guaranteeed → guaranteed)

maribu committed:
Merge pull request #11998 from gschorcht/cpu/atmega_common/malloc-wrappers

cpu/atmega_common: wrappers for memory management function to avoid preemption

maribu committed:
Merge pull request #10555 from gschorcht/irq_handler

sys: single interrupt handler thread for interrupts in modules with blocking functions

smlng committed:
Merge pull request #10252 from jcarrano/flake8-report-existence

tools/ci: correcly report flake8 version.

haukepetersen committed:
cpu/nrf5x: fix bitrate doc for nrfmin driver

cladmi committed:
makefiles/buildtest: add a buildtest-indocker

Add a 'buildtest-indocker' that forces executing 'buildtest' for loop
completely inside the container.
It prevents starting one container per compilation wich is slower but
it could hide errors where the host toolchain would be used

It is currently equivalent to `buildtest` but will change when the
`buidtest` handling will be move outside of `BUILD_IN_DOCKER`.

Display an error when executed without BUILD_IN_DOCKER=1.

fjmolinas committed:
Merge pull request #11083 from cladmi/pr/docker/hack/bootloader

 tests/mcuboot: handle building in docker

fjmolinas committed:
Merge pull request #11857 from cladmi/pr/make/buildtest/docker

makefiles/buildtest: always execute 'buildtest' loop on host machine

jcarrano committed:
tests/shell: Use socat as the RIOT_TERMINAL.

For test scripts, a terminal that does not modify the input and output
streams, configured without local echo, is preferred as it ensures the
test setup is introducing as little noise as possible.

fjmolinas committed:
Merge pull request #11875 from cladmi/pr/test_utils/sync

test_utils_interactive_sync: add a helper for synchronizing tests

smlng committed:
Merge pull request #9852 from bergzand/pr/rpl/fix_debug_hdrs

gnrc_rpl: remove DEBUG from header files

kaspar030 committed:
Merge pull request #11004 from jcarrano/shell_cancel_line_ctrl-c

 sys/shell: cancel current line on CTRL-C.

MrKevinWeiss committed:
cpu/stm32: Fix 16 bit reg endianess for i2c_1

Since i2c_1 i2c_write_regs is cpu specific reg endianess must be swapped there
Change reg from little endian to big endian

maribu committed:
Merge pull request #12023 from MrKevinWeiss/pr/stm/i2c1/endianess

cpu/stm32: Fix 16 bit reg endianness for i2c_1

gschorcht committed:
cpu/esp32: fixes printf and puts

printf and puts used ets_printf before. Unfortunately, ets_printf adds an additional \r for each \n which is not consistent with other RIOT platforms. As a result some automatic tests failed. Therefore, both functions write now character-wise directly to the UART interface.

Commit:
Merge pull request #11991 from gschorcht/cpu/stp32_common/can/auto-init-cleanup

sys/auto_init/can: cleanup for STM32 CAN controller driver

kaspar030 committed:
Merge pull request #12026 from gschorcht/cpu/esp32/printf_puts_fix

cpu/esp32: fixes printf and puts

gschorcht committed:
cpu/esp32: use always newlib_syscalls_default

Modules newlib and newlib_syscalls_default are now used by default. Conditional compilations for MODULE_NEWLIB_SYSCALLS_DEFAULT as well as alternative code are removed completely.

benpicco committed:
Merge pull request #12001 from nmeum/pr/gnrc_tcp_memory_leak

gnrc_tcp: Fix memory leak, potential DOS

gschorcht committed:
cpu/esp32: use printf/puts from newlib

Initializing the stdio file descriptors in global reent structure with newlib fake stdio file descriptors led to the problem that newlib stdio functions printf and puts were not working since they can't operate on these fake stdio file descriptors. Therefore, this initialization was removed. Now, the real stdio file descriptors as created automatically by newlib are used. Specific functions `printf`, `puts`, `getchar`and `putchar` are not required any longer and are removed now.

kaspar030 committed:
Merge pull request #12029 from gschorcht/cpu/esp32/newlib_fix

cpu/esp32: fix newlib usage for stdio functions

PeterKietzmann committed:
Merge pull request #12005 from haukepetersen/fix_nrfmin_doc

cpu/nrf5x: fix bitrate doc for nrfmin driver

kaspar030 committed:
fixup! examples/suit_update: add SUIT draft v4 example & test

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! fixup! examples/suit_update: add SUIT draft v4 example & test

kaspar030 committed:
fixup! fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 committed:
fixup! fixup! sys/suit: add SUIT draft v4 firmware upgrade module

kaspar030 commented:
Closing, as this is PRed to mainline.
